// @algorandfoundation/algokit-client-generator/schema/contract.schema.json.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var $schema = "http://json-schema.org/draft-07/schema";
var title = "AbiContract";
var type = "object";
var required = [
	"name",
	"methods"
];
var additionalProperties = true;
var properties = {
	name: {
		type: "string"
	},
	desc: {
		type: "string"
	},
	methods: {
		type: "array",
		items: {
			$ref: "#/definitions/ContractMethod"
		}
	},
	events: {
		type: "array",
		items: {
			$ref: "#/definitions/Event"
		}
	},
	networks: {
		type: "object",
		additionalProperties: {
			type: "object",
			required: [
				"appID"
			],
			additionalProperties: true,
			properties: {
				appID: {
					type: "number"
				}
			}
		}
	}
};
var definitions = {
	ContractMethod: {
		type: "object",
		additionalProperties: true,
		required: [
			"name",
			"args",
			"returns"
		],
		properties: {
			name: {
				type: "string"
			},
			args: {
				type: "array",
				items: {
					$ref: "#/definitions/ContractMethodArg"
				}
			},
			desc: {
				type: "string"
			},
			returns: {
				type: "object",
				additionalProperties: true,
				required: [
					"type"
				],
				properties: {
					desc: {
						type: "string"
					},
					type: {
						$ref: "#/definitions/ABIType"
					}
				}
			},
			readonly: {
				type: "boolean"
			},
			events: {
				type: "array",
				items: {
					$ref: "#/definitions/Event"
				}
			}
		}
	},
	Event: {
		type: "object",
		additionalProperties: true,
		required: [
			"name",
			"args"
		],
		properties: {
			name: {
				type: "string"
			},
			args: {
				type: "array",
				items: {
					$ref: "#/definitions/ContractMethodArg"
				}
			},
			desc: {
				type: "string"
			}
		}
	},
	ContractMethodArg: {
		type: "object",
		additionalProperties: true,
		required: [
			"name",
			"type"
		],
		properties: {
			desc: {
				type: "string"
			},
			type: {
				$ref: "#/definitions/ABIType"
			},
			name: {
				type: "string"
			}
		}
	},
	ABIType: {
		description: "Catch all for fixed length arrays and tuples",
		type: "string"
	}
};
var contractSchema = {
	$schema: $schema,
	title: title,
	type: type,
	required: required,
	additionalProperties: additionalProperties,
	properties: properties,
	definitions: definitions
};

exports.$schema = $schema;
exports.additionalProperties = additionalProperties;
exports.default = contractSchema;
exports.definitions = definitions;
exports.properties = properties;
exports.required = required;
exports.title = title;
exports.type = type;
