// @algorandfoundation/algokit-client-generator/client/generate.mjs
import { inline, NewLine } from '../output/writer.mjs';
import { paramsFactory } from './params-factory.mjs';
import { appClient } from './app-client.mjs';
import { deployTypes } from './deploy-types.mjs';
import { utilityTypes } from './utility-types.mjs';
import { imports } from './imports.mjs';
import { createGeneratorContext } from './generator-context.mjs';
import { appTypes } from './app-types.mjs';
import { callComposerType } from './call-composer-types.mjs';
import { appFactory } from './app-factory.mjs';

function convertStructs(s, sanitizer) {
    return s.map(({ name, type }) => ({
        name: sanitizer.makeSafePropertyIdentifier(name),
        type: typeof type === 'string' ? type : convertStructs(type, sanitizer),
    }));
}
function shrinkAppSpec(app, options) {
    const strippedAppSpec = structuredClone(app);
    // Only keep the source info if it is needed for error mapping
    const shrinkSourceInfo = (sourceInfo) => {
        return sourceInfo
            .filter((entry) => entry.errorMessage)
            .map((entry) => ({
            pc: entry.pc,
            errorMessage: entry.errorMessage,
            // Keep minimal context for error mapping if available
            ...(entry.teal !== undefined && { teal: entry.teal }),
        }));
    };
    // Keep only source info entries that can be used for approval and clear program error mapping
    if (strippedAppSpec.sourceInfo?.approval?.sourceInfo && strippedAppSpec.sourceInfo.approval.sourceInfo.length > 0) {
        strippedAppSpec.sourceInfo.approval.sourceInfo = shrinkSourceInfo(strippedAppSpec.sourceInfo.approval.sourceInfo);
    }
    if (strippedAppSpec.sourceInfo?.clear?.sourceInfo && strippedAppSpec.sourceInfo.clear.sourceInfo.length > 0) {
        strippedAppSpec.sourceInfo.clear.sourceInfo = shrinkSourceInfo(strippedAppSpec.sourceInfo.clear.sourceInfo);
    }
    if (strippedAppSpec.compilerInfo) {
        delete strippedAppSpec.compilerInfo;
    }
    // These are used for deploying but not for calling deployed apps
    if (options.mode === 'minimal') {
        if (strippedAppSpec.source) {
            delete strippedAppSpec.source;
        }
        if (strippedAppSpec.byteCode) {
            delete strippedAppSpec.byteCode;
        }
        if (strippedAppSpec.templateVariables) {
            delete strippedAppSpec.templateVariables;
        }
        if (strippedAppSpec.scratchVariables) {
            delete strippedAppSpec.scratchVariables;
        }
    }
    return strippedAppSpec;
}
function* generate(app, options) {
    const resolvedOptions = {
        // Set defaults
        preserveNames: false,
        mode: 'full',
        ...options,
    };
    const reduceAppSpec = shrinkAppSpec(app, resolvedOptions);
    const ctx = createGeneratorContext(reduceAppSpec, resolvedOptions);
    yield `/* eslint-disable */`;
    yield `/**`;
    yield ` * This file was automatically generated by @algorandfoundation/algokit-client-generator.`;
    yield ` * DO NOT MODIFY IT BY HAND.`;
    yield ` * requires: @algorandfoundation/algokit-utils: ^7`;
    yield ` */`;
    yield* imports(ctx);
    // Change the structs definition to sanitize property names according to the defined rules
    // for instance, this may (unless you passed in --preserve-names) convert properties like my_prop to myProp
    reduceAppSpec.structs = Object.fromEntries(Object.keys(reduceAppSpec.structs).map((key) => [key, convertStructs(reduceAppSpec.structs[key], ctx.sanitizer)]));
    yield* inline('export const APP_SPEC: Arc56Contract = ', JSON.stringify(reduceAppSpec), ' as unknown as Arc56Contract');
    yield NewLine;
    yield* utilityTypes();
    yield NewLine;
    yield* appTypes(ctx);
    if (ctx.mode === 'full') {
        yield* deployTypes(ctx);
    }
    yield NewLine;
    // Write a call factory
    yield* paramsFactory(ctx);
    yield NewLine;
    // Write a factory in full mode
    if (ctx.mode === 'full') {
        yield* appFactory(ctx);
    }
    // Write a client
    yield* appClient(ctx);
    yield* callComposerType(ctx);
}

export { generate };
